src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -176,62 +190,79 @@ internal static unsafe bool IsValidCanonical(char* name, int start, ref int end,
         // Return Invalid (-1) for failures.
         // If the address has less than three dots, only the rightmost section is assumed to contain the combined value for
         // the missing sections: 0xFF00FFFF == 0xFF.0x00.0xFF.0xFF == 0xFF.0xFFFF
-        internal static unsafe long ParseNonCanonical(char* name, int start, ref int end, bool notImplicitFile)
+        internal static long ParseNonCanonical<TChar>(in ReadOnlySpan<TChar> name, out int charsConsumed, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
-            int numberBase = Decimal;
-            char ch;
-            long* parts = stackalloc long[4];
+            int numberBase = IPv4AddressHelper.Decimal;
+            Span<uint> parts = stackalloc uint[4];
```

Comment:
This is an example where you're going to be incurring bounds checks where the code didn't previously.

I'd encourage you to reduce this PR to just the minimum required to implement the feature, which presumably means e.g. changing `char* name` to `TChar* name` rather than introducing spans. A separate PR can then focus just on reducing the use of unsafe code.

Reply:
Will do. I thought the `for` loop would eliminate those bounds checks (and I didn't see any in the IL) but I've checked the ASM generated by the JIT and can see them. Strange!